File: src/jvmch/JavaCompiler.java
package jvmch;

import javax.tools.ToolProvider;
import java.io.*;
import java.util.*;
import java.util.jar.*;

/**
 * Created by Niyaz Nigmatullin on 11.04.17.
 * Ported to Java by Maxim Buzdalov on 03.09.17.
 */
public class JavaCompiler extends Module {
    private static final int COPY_BUFFER_SIZE = 0x10000;
    private static final Random rng = new Random(239L);

    @Override
    public boolean checkArgs(String[] args, int argumentOffset) {
        return args.length >= 3;
    }

    @Override
    public boolean run(String[] args) {
        File tempDir = new File(args[0]);
        File jarFile = new File(args[1]);
        File[] sourceFiles = new File[args.length - 2];
        for (int i = 0; i < sourceFiles.length; ++i) {
            sourceFiles[i] = new File(args[i + 2]);
        }
        int exitCode = compile(tempDir, jarFile, sourceFiles);
        return exitCode == 0;
    }

    @Override
    public String getUsage() {
        return "compiles Java source files given. Arguments:\n"
                + "        <temporary directory> <resulting jar file> <source files>";
    }

    private int compile(File tempDir, File file, File[] sources) {
        try {
            int exitCode = compile(tempDir, copyFiles(tempDir, sources));
            if (exitCode != 0) {
                return exitCode;
            }
            List<String> classes = new ArrayList<>();
            findFiles(tempDir, "", ".class", classes);
            createJar(file, tempDir, classes.toArray(new String[classes.size()]));
            return 0;
        } catch (IOException e) {
            e.printStackTrace();
            return 100;
        }
    }

    private int compile(File dir, File[] sources) throws IOException {
        String[] args = new String[sources.length + 2];
        args[0] = "-d";
        args[1] = dir.getCanonicalPath();
        for (int i = 0; i < sources.length; ++i) {
            args[i + 2] = sources[i].getCanonicalPath();
        }
        return ToolProvider.getSystemJavaCompiler().run(null, null, null, args);
    }

    private File[] copyFiles(File dir, File[] files) throws IOException {
        dir.mkdirs();
        File[] rv = new File[files.length];
        for (int i = 0; i < files.length; ++i) {
            rv[i] = new File(dir, getFilename(files[i]));
            copyFile(files[i], rv[i]);
        }
        return rv;
    }

    private String getFilename(File file) throws IOException {
        return getClassName(file).replace('.', '/') + ".java";
    }

    private void findFiles(File dir, String path, String extension, List<String> result) {
        File[] files = dir.listFiles();
        if (files != null) {
            for (File file : files) {
                if (file.isDirectory()) {
                    findFiles(file, path + file.getName() + "/", extension, result);
                } else if (file.getName().endsWith(extension)) {
                    result.add(path + file.getName());
                }
            }
        }
    }

    private String getClassName(File file) throws IOException {
        try (FileReader reader = new FileReader(file)) {
            StreamTokenizer tokenizer = new StreamTokenizer(reader);
            tokenizer.wordChars('_', '_');
            tokenizer.wordChars('0', '9');
            tokenizer.wordChars('$', '$');
            tokenizer.slashSlashComments(true);
            tokenizer.slashStarComments(true);
            String packageName = "";
            int blocks = 0;
            boolean publicFound = false;
            while (tokenizer.ttype != StreamTokenizer.TT_EOF) {
                if (tokenizer.ttype == '{') {
                    blocks++;
                } else if (tokenizer.ttype == '}') {
                    blocks--;
                } else if (tokenizer.ttype == StreamTokenizer.TT_WORD) {
                    if ("package".equals(tokenizer.sval)) {
                        do {
                            tokenizer.nextToken();
                            packageName += tokenizer.sval + '.';
                            tokenizer.nextToken();
                        } while (tokenizer.ttype != ';');
                    } else if ("public".equals(tokenizer.sval) && blocks == 0) {
                        publicFound = true;
                    } else if (("class".equals(tokenizer.sval) || "interface".equals(tokenizer.sval)) && publicFound) {
                        tokenizer.nextToken();
                        return packageName + tokenizer.sval;
                    }
                }
                tokenizer.nextToken();
            }
            StringBuilder randomName = new StringBuilder();
            for (int i = 0; i < 10; ++i) {
                randomName.append((char) (rng.nextInt(26) + 'a'));
            }
            return packageName + "$$__" + randomName.toString();
        }
    }

    private void copyStream(InputStream inputStream, OutputStream outputStream) throws IOException {
        byte[] buffer = new byte[COPY_BUFFER_SIZE];
        int bytes;
        while ((bytes = inputStream.read(buffer)) > 0) {
            outputStream.write(buffer, 0, bytes);
        }
    }

    private void copyFile(File fromFile, File toFile) throws IOException {
        toFile.getAbsoluteFile().getParentFile().mkdirs();
        try (BufferedInputStream inputStream = new BufferedInputStream(new FileInputStream(fromFile));
             BufferedOutputStream outputStream = new BufferedOutputStream(new FileOutputStream(toFile))) {
            copyStream(inputStream, outputStream);
        }
    }

    private void createJar(File jarFile, File dir, String[] files) throws IOException {
        try (FileOutputStream fos = new FileOutputStream(jarFile);
             JarOutputStream jos = new JarOutputStream(fos)) {
            for (String file : files) {
                jos.putNextEntry(new JarEntry(file));
                try (FileInputStream fis = new FileInputStream(new File(dir, file));
                     BufferedInputStream inputStream = new BufferedInputStream(fis)) {
                    copyStream(inputStream, jos);
                }
            }
        }
    }
}

File: src/jvmch/JavaCompiler.kt
package jvmch

import javax.tools.ToolProvider
import java.io.*
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarOutputStream

/**
 * Created by Niyaz Nigmatullin on 11.04.17.
 */
class JavaCompiler : Module() {
    private val COPY_BUFFER_SIZE = 0x10000
    private val rng = Random(239L)

    override fun checkArgs(args: Array<String>, argumentOffset: Int): Boolean {
        return args.size >= 3
    }

    override fun run(args: Array<String>): Boolean {
        val tempDir = File(args[0])
        val jarFile = File(args[1])
        val sourceFiles = args.takeLast(args.size - 2).map(::File).toTypedArray()
        val exitCode = compile(tempDir, jarFile, sourceFiles)
        return exitCode == 0
    }

    override fun getUsage(): String {
        return "compiles Java source files given. Arguments:\n" +
                "        <temporary directory> <resulting jar file> <source files>"
    }

    private fun compile(tempDir: File, file: File, sources: Array<File>): Int {
        try {
            val exitCode = compile(tempDir, copyFiles(tempDir, sources))
            if (exitCode != 0) {
                return exitCode
            }
            val classes = mutableListOf<String>()
            findFiles(tempDir, "", ".class", classes)
            createJar(file, tempDir, classes.toTypedArray())
            return 0
        } catch (e: IOException) {
            e.printStackTrace()
            return 100
        }
    }

    private fun compile(dir: File, sources: Array<File>): Int {
        val args = arrayOfNulls<String>(sources.size + 2)
        args[0] = "-d"
        args[1] = dir.canonicalPath
        for (i in sources.indices) {
            args[i + 2] = sources[i].canonicalPath
        }
        val compiler = ToolProvider.getSystemJavaCompiler()
        return compiler.run(null, null, null, *args)
    }

    private fun copyFiles(dir: File, files: Array<File>): Array<File> {
        dir.mkdirs()
        return files.map {
            val file = File(dir, getFilename(it))
            copyFile(it, file)
            file
        }.toTypedArray()
    }

    private fun getFilename(file: File): String = getClassName(file).replace('.', '/') + ".java"

    private fun findFiles(dir: File, path: String, extension: String, result: MutableCollection<String>) {
        val files = dir.listFiles() ?: return
        for (file in files) {
            if (file.isDirectory) {
                findFiles(file, path + file.name + "/", extension, result)
            } else if (file.name.endsWith(extension)) {
                result.add(path + file.name)
            }
        }
    }

    private fun getClassName(file: File): String {
        FileReader(file).use { reader ->
            val tokenizer = StreamTokenizer(reader)
            tokenizer.wordChars('_'.toInt(), '_'.toInt())
            tokenizer.wordChars('0'.toInt(), '9'.toInt())
            tokenizer.wordChars('$'.toInt(), '$'.toInt())
            tokenizer.slashSlashComments(true)
            tokenizer.slashStarComments(true)
            var packageName = ""
            var blocks = 0
            var publicFound = false
            while (tokenizer.ttype != StreamTokenizer.TT_EOF) {
                if (tokenizer.ttype == '{'.toInt()) {
                    blocks++
                } else if (tokenizer.ttype == '}'.toInt()) {
                    blocks--
                } else if (tokenizer.ttype == StreamTokenizer.TT_WORD) {
                    if ("package" == tokenizer.sval) {
                        do {
                            tokenizer.nextToken()
                            packageName = tokenizer.sval + '.'
                            tokenizer.nextToken()
                        } while (tokenizer.ttype != ';'.toInt())
                    } else if ("public" == tokenizer.sval && blocks == 0) {
                        publicFound = true
                    } else if (("class" == tokenizer.sval || "interface" == tokenizer.sval) && publicFound) {
                        tokenizer.nextToken()
                        return packageName + tokenizer.sval
                    }
                }
                tokenizer.nextToken()
            }
            var randomName = ""
            for (i in 0..10) {
                randomName += (rng.nextInt(26) + 'a'.toInt()).toChar()
            }
            return packageName + "\$\$__" + randomName
        }
    }

    private fun copyStream(inputStream: InputStream, outputStream: OutputStream) {
        val buffer = ByteArray(COPY_BUFFER_SIZE)
        while (true) {
            val bytes = inputStream.read(buffer)
            if (bytes == -1) break
            outputStream.write(buffer, 0, bytes)
        }
    }

    private fun copyFile(fromFile: File, toFile: File) {
        toFile.absoluteFile.parentFile.mkdirs()
        BufferedInputStream(FileInputStream(fromFile)).use { inputStream ->
            BufferedOutputStream(FileOutputStream(toFile)).use { outputStream ->
                copyStream(inputStream, outputStream)
            }
        }
    }

    private fun createJar(jarFile: File, dir: File, files: Array<String>) {
        JarOutputStream(FileOutputStream(jarFile)).use { jos ->
            for (file in files) {
                jos.putNextEntry(JarEntry(file))
                BufferedInputStream(FileInputStream(File(dir, file))).use { inputStream ->
                    copyStream(inputStream, jos)
                }
            }
        }
    }
}
